pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentReducer.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.pinot.core.segment.processing.collector.Collector;
+import org.apache.pinot.core.segment.processing.collector.CollectorFactory;
+import org.apache.pinot.core.segment.processing.utils.SegmentProcessorUtils;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.data.readers.RecordReader;
+import org.apache.pinot.spi.data.readers.RecordReaderFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Reducer phase of the SegmentProcessorFramework
+ * Reads the avro files in the input directory and creates output avro files in the reducer output directory.
+ * The avro files in the input directory are expected to contain data for only 1 partition
+ * Performs operations on that partition data as follows:
+ * - concatenation/rollup of records
+ * - split
+ * - TODO: dedup
+ */
+public class SegmentReducer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentReducer.class);
+  private static final int MAX_RECORDS_TO_COLLECT = 5_000_000;
+
+  private final File _reducerInputDir;
+  private final File _reducerOutputDir;
+
+  private final String _reducerId;
+  private final Schema _pinotSchema;
+  private final org.apache.avro.Schema _avroSchema;
+  private final Collector _collector;
+  private final int _numRecordsPerPart;
+
+  public SegmentReducer(String reducerId, File reducerInputDir, SegmentReducerConfig reducerConfig,
+      File reducerOutputDir) {
+    _reducerInputDir = reducerInputDir;
+    _reducerOutputDir = reducerOutputDir;
+
+    _reducerId = reducerId;
+    _pinotSchema = reducerConfig.getPinotSchema();
+    _avroSchema = SegmentProcessorUtils.convertPinotSchemaToAvroSchema(_pinotSchema);
+    _collector = CollectorFactory.getCollector(reducerConfig.getCollectorConfig(), _pinotSchema);
+    _numRecordsPerPart = reducerConfig.getNumRecordsPerPart();
+    LOGGER.info("Initialized reducer with id: {}, input dir: {}, output dir: {}, collector: {}, numRecordsPerPart: {}",
+        _reducerId, _reducerInputDir, _reducerOutputDir, _collector.getClass(), _numRecordsPerPart);
+  }
+
+  /**
+   * Reads the avro files in the input directory.
+   * Performs configured operations and outputs to other avro file(s) in the reducer output directory.
+   */
+  public void reduce()
+      throws Exception {
+
+    int part = 0;
+    for (File inputFile : _reducerInputDir.listFiles()) {
+
+      RecordReader avroRecordReader = RecordReaderFactory
+          .getRecordReaderByClass("org.apache.pinot.plugin.inputformat.avro.AvroRecordReader", inputFile,
+              _pinotSchema.getColumnNames(), null);
+
+      while (avroRecordReader.hasNext()) {
+        GenericRow next = avroRecordReader.next();
+
+        // Aggregations
+        _collector.collect(next);
+
+        // Exceeded max records allowed to collect. Flush
+        if (_collector.size() == MAX_RECORDS_TO_COLLECT) {
```

Comment:
Should we use `_numRecordsPerPart` here so that once the collector collects enough records, we flush them?

Reply:
Instead of checking the number of records, can we have a method in the collector that says it is full? The criteria could then be something else depending on the collector.

Reply:
`Should we use _numRecordsPerPart here so that once the collector collects enough records, we flush them` - I had initially done it this way. But then I noticed while testing, that if user sets a very low numRecordsPerPart, we will aggregate less and flush very frequently. Having a bigger MAX_RECORDS_TO_COLLECT allows for more aggregation

Reply:
I think it is fine if user configures a very low `_numRecordsPerPart` and not get much aggregation. The problem of using the fixed large threshold is that for use cases with lots of columns, we might run out of memory and there is no knob to tune it.
In order to get better aggregation, we can potentially do an external sort on the files before collecting the records. Not requires in the first version.

Reply:
sounds good. Changed it to flush when numRecordsPerPart is reached.
@mcvsubbu we can certainly do that if we start seeing need for more advanced controls and tuning on collection size and flushing. For starters, I wanted to keep it simple